### PR TITLE
Add 'MeanTimeReporter';

### PR DESCRIPTION
**Pull request by gavinlaking**
Using a temporary 'previous runs' file, we gather statistics on previous
test run times. From this file we build a report which indicates the average
test run time, a minimum (fastest) test run time, a maximum (slowest) test
run time for each test. A description is included for each report line which
is the test name as reported by Minitest.

Tests to follow, pending #173.

### CODE REVIEW

**Comment by os97673** [comment]
Thank you for the contribution.  Just a few comments:
- it would be nice to have some description of the reporter (as a class comment)
- could you please add the reporter to the list of reporters which are executed as part of `rake gallery`

File: lib/minitest/reporters/mean_time_reporter.rb [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+      #   the maximum time the test took to run and a description of the test
+      #   (which is the test name as emitted by Minitest).
+      def report_filename
+        options[:report_filename]
+      end
+
+      # A barbaric way to find out how many runs are in the previous runs file;
+      # this method takes the first test listed, and counts its samples
+      # trusting (naively) all runs to be the same number of samples. This will
+      # produce incorrect averages when new tests are added, so it is advised
```
Comment (os97673) [inline lib/minitest/reporters/mean_time_reporter.rb]: Hmm, I'm not sure I like the fact that it is user who is responsible for resetting the data.
Perhaps we could change this and either reset it for her or be able to report stats for both old and new tests.
WDYT?

**Comment by gavinlaking** [comment]
Thanks for the feedback! I'll make the necessary changes.

One consideration which has arisen is how to handle CI. At the moment, TravisCI will not allow you to write to a `/tmp` directory (as there isn't one) to run this reporter, so I'm not sure how to proceed there.

I've also cooked up a `rake` task to reset the statistics file, which reduces some of the burden of resetting the data. I expect that this reporter will be used for many 'runs'. I use this for a project with ~2500 tests and after a couple of hours there are quite a few samples in there. It would be nice to preserve these as a feature could takes days/weeks to build and I might want to see regressions over time. Again, an area that might need a bit of attention.

**Comment by gavinlaking** [comment]
Hi Oleg,

Hope you are well. Do you have any suggestions for how I might demonstrate the functionality of this pull request via tests which are useful to you to prevent regressions and increase the likelihood of a merge?

Gav

**Comment by os97673** [comment]
Actually, with new changes the reporter looks fine for me.  Let's see what users will say ;)

**Pull request merged**